### PR TITLE
Remove leftover empty files when closing the plugin.

### PR DESCRIPTION
**Pull request by mashhurs**
## Release notes

[rn:skip]

## What does this PR do?
When closing the plugin, the plugin checks if file needs to be uploaded to S3. If file is empty, it will be removed (without this change the file will be left).

## Why is it important/What is the impact to the user?
For intensive S3 upload operations and rotation policy is fast (size or time), multiple files will be created. When the plugin is being closed, created but not written to files will be just left. There will be many empty files over time.
Note that, we introduced separate logic to remove the empty dirs but it doesn't remove if dir has files, even empty files.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] ~~I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## How to test this PR locally
- use the following config
```
input {
  generator {
        id => "generator-id"
        ecs_compatibility => disabled
        #enable_metric => true
        count => 10000
        threads => 10
        codec => json
        lines => [
            '{"fileset":{"module":"system","name":"asa"},"system":{"auth":{"timestamp":"May 17  05:17:00","ssh":{"source":{"ip":"54.160.29.55"}}}},"event":{"category":"cisco-category", "type":"cisco-type", "data":{"User-Name":"mashhur"}},"client":{"ar_net":"54.160.29.55", "ongisac_ip":"54.160.29.55", "ip":"54.160.29.55"}, "destination": {"ar_net":"54.160.29.55", "ongisac_ip":"54.160.29.55"}, "source": {"ar_net":"54.160.29.55", "ongisac_ip":"54.160.29.55"}, "url":{"ongisac_domain": "ip-172-31-4-132"}, "DstIP":"54.160.29.55","SrcIP":"54.160.29.55","orginalClientSrcIP":"54.160.29.55","ReferencedHost":"ip-172-31-4-132", "dns":{"question": {"ongisac_domain":"example.com/my-path?query=value"}}}'
        ]
      }
}
output {
    s3 {
        access_key_id => "{yourAccessID}"
        secret_access_key => "{yourAccessKey}"
        region => "aws-region-n"
        bucket => "buckjet-name"
        codec => "json_lines"
        canned_acl => "private"
        prefix => "test-%{+YYYY.MM.dd}"
        additional_settings => {
            "force_path_style" => true
        }
        upload_queue_size => 10
        upload_workers_count => 10
        rotation_strategy => "size_and_time"
	size_file => 500
	time_file => 1
        temporary_directory => "/elastic/s3-output/temp"
        validate_credentials_on_root_bucket => false
    }
}
```
- run the Logstash (`./bin/logstash -f your-config-file.conf`)
- check the `/elastic/s3-output/temp` temp dir that no any files left after stopping Logstash. Without this change, created but not written files will be left.

## Related issues
- 

## Logs

```
PS C:\ls> .\bin\logstash -f .\config\s3.config.txt
"Using bundled JDK: C:\ls\jdk\bin\java.exe"
Sending Logstash logs to C:/ls/logs which is now configured via log4j2.properties
[2024-06-21T22:16:00,584][INFO ][logstash.runner          ] Log4j configuration path used is: C:\ls\config\log4j2.properties
[2024-06-21T22:16:00,597][INFO ][logstash.runner          ] Starting Logstash {"logstash.version"=>"8.15.0", "jruby.version"=>"jruby 9.4.7.0 (3.1.4) 2024-04-29 597ff08ac1 OpenJDK 64-Bit Server VM 21.0.3+9-LTS on 21.0.3+9-LTS +indy +jit [x86_64-mswin32]"}
[2024-06-21T22:16:00,597][INFO ][logstash.runner          ] JVM bootstrap flags: [-Xms1g, -Xmx1g, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djruby.compile.invokedynamic=true, -XX:+HeapDumpOnOutOfMemoryError, -Djava.security.egd=file:/dev/urandom, -Dlog4j2.isThreadContextMapInheritable=true, -Dlogstash.jackson.stream-read-constraints.max-string-length=200000000, -Dlogstash.jackson.stream-read-constraints.max-number-length=10000, -Djruby.regexp.interruptible=true, -Djdk.io.File.enableADS=true, --add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED, --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED, --add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED, --add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED, --add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED, --add-opens=java.base/java.security=ALL-UNNAMED, --add-opens=java.base/java.io=ALL-UNNAMED, --add-opens=java.base/java.nio.channels=ALL-UNNAMED, --add-opens=java.base/sun.nio.ch=ALL-UNNAMED, --add-opens=java.management/sun.management=ALL-UNNAMED, -Dio.netty.allocator.maxOrder=11]
[2024-06-21T22:16:00,597][INFO ][logstash.runner          ] Jackson default value override `logstash.jackson.stream-read-constraints.max-string-length` configured to `200000000`
[2024-06-21T22:16:00,597][INFO ][logstash.runner          ] Jackson default value override `logstash.jackson.stream-read-constraints.max-number-length` configured to `10000`
[2024-06-21T22:16:00,678][WARN ][logstash.config.source.multilocal] Ignoring the 'pipelines.yml' file because modules or command line options are specified
[2024-06-21T22:16:02,119][INFO ][logstash.agent           ] Successfully started Logstash API endpoint {:port=>9600, :ssl_enabled=>false}
[2024-06-21T22:16:02,913][INFO ][org.reflections.Reflections] Reflections took 175 ms to scan 1 urls, producing 138 keys and 481 values
[2024-06-21T22:16:03,062][INFO ][logstash.codecs.json     ] ECS compatibility is enabled but `target` option was not specified. This may cause fields to be set at the top-level of the event where they are likely to clash with the Elastic Common Schema. It is recommended to set the `target` option to avoid potential schema conflicts (if your data is ECS compliant or non-conflicting, feel free to ignore this message)
[2024-06-21T22:16:06,613][INFO ][logstash.codecs.jsonlines] ECS compatibility is enabled but `target` option was not specified. This may cause fields to be set at the top-level of the event where they are likely to clash with the Elastic Common Schema. It is recommended to set the `target` option to avoid potential schema conflicts (if your data is ECS compliant or non-conflicting, feel free to ignore this message)
[2024-06-21T22:16:08,726][INFO ][logstash.javapipeline    ] Pipeline `main` is configured with `pipeline.ecs_compatibility: v8` setting. All plugins in this pipeline will default to `ecs_compatibility => v8` unless explicitly configured otherwise.
[2024-06-21T22:16:09,025][INFO ][logstash.javapipeline    ][main] Starting pipeline {:pipeline_id=>"main", "pipeline.workers"=>8, "pipeline.batch.size"=>125, "pipeline.batch.delay"=>50, "pipeline.max_inflight"=>1000, "pipeline.sources"=>["C:/ls/config/s3.config.txt"], :thread=>"#<Thread:0x4f976cae C:/ls/logstash-core/lib/logstash/java_pipeline.rb:134 run>"}
[2024-06-21T22:16:10,164][INFO ][logstash.javapipeline    ][main] Pipeline Java execution initialization time {"seconds"=>1.14}
[2024-06-21T22:16:10,171][INFO ][logstash.javapipeline    ][main] Pipeline started {"pipeline.id"=>"main"}
[2024-06-21T22:16:10,190][INFO ][logstash.agent           ] Pipelines running {:count=>1, :running_pipelines=>[:main], :non_running_pipelines=>[]}
[2024-06-21T22:16:15,688][INFO ][logstash.javapipeline    ][main] Pipeline terminated {"pipeline.id"=>"main"}
[2024-06-21T22:16:15,848][INFO ][logstash.pipelinesregistry] Removed pipeline from registry successfully {:pipeline_id=>:main}
[2024-06-21T22:16:15,860][INFO ][logstash.runner          ] Logstash shut down.
PS C:\ls>
```

### CODE REVIEW

**Comment by mashhurs** [comment]
Failed CI will be 🟢 with 8.14.2 soon. https://github.com/elastic/logstash/pull/16255 unit test is a safeguard for such failure cases.

**Review by edmocosta — COMMENTED** [review]
LGTM, just minor suggestions: 
- How about adding a `debug` log [here](https://github.com/logstash-plugins/logstash-integration-aws/pull/46/files#diff-bdae3d005e86be10100b9ba58d128b55d13c8b8bb4e4ffd43f2c0956c22666e2R376), just in case we need troubleshooting those temporary file deletions.
- Could we add a test to ensure those empty files are deleted when the `close` is called?

**Comment by mashhurs** [comment]
> LGTM, just minor suggestions:
> 
> * How about adding a `debug` log [here](https://github.com/logstash-plugins/logstash-integration-aws/pull/46/files#diff-bdae3d005e86be10100b9ba58d128b55d13c8b8bb4e4ffd43f2c0956c22666e2R376), just in case we need troubleshooting those temporary file deletions.

I once more thought and moved the removal logic in the `plugin#close`. In `clean_temporary_file` there are debugging log lines and LS core has also debugging log before closing the plugin as I remember.

> * Could we add a test to ensure those empty files are deleted when the `close` is called?

Nice, I have just added it.

File: lib/logstash/outputs/s3.rb [hunk 1]
```diff
@@ -272,6 +272,7 @@ def close
     # This will block the shutdown until all upload are done or the use force quit.
     @file_repository.each_files do |file|
       upload_file(file)
+      clean_temporary_file(file) if file.size == 0
```
Comment (edmocosta) [inline lib/logstash/outputs/s3.rb:275]: > I once more thought and moved the removal logic in the plugin#close. In clean_temporary_file there are debugging log lines and LS core has also debugging log before closing the plugin as I remember.

Agreed, I think it makes sense to move this logic here, as it seems the `close` method is the only place that might left empty files undeleted.

I've a concern about the latest change. If I'm not mistaken, not all `@file_repository.each_files` will be empty files, meaning that they might be deleted by the `upload_file` method before reaching this size check (`file.size == 0`), raising an error.

I haven't tested it, but I think an option would be using the `File.zero?`, or moving this logic after the uploader was stopped.

```suggestion
      clean_temporary_file(file) if File.zero?(file.key)
```

I guess It wasn't caught by the added tests because the files are not being synchronously deleted and/or timing of executing the `:on_complete` callback.

File: spec/outputs/s3_spec.rb [hunk 2]
```diff
@@ -258,4 +258,25 @@
...
+    end
+
+    it "removes empty directory" do
+      subject.register
+      expect(Dir.exist?(temporary_directory)).to be_truthy
+
+      subject.multi_receive_encoded(events_and_encoded) # creates a temporary dir
+      expect((Dir.entries(temporary_directory) - %w{ . .. }).empty?).to be_falsey
+
+      sleep(20) # let file rotation happen and leave new empty dir
```
Comment (edmocosta) [inline spec/outputs/s3_spec.rb:276]:
```suggestion
      subject.instance_variable_get(:@periodic_check).execute
```
I'm not sure if this suggestion is a Ruby's bad practice, but I think we could force the rotate here instead of waiting 20 secs. Other option would be extracting the @periodic_check body code into a method, and just invoke it :)

File: lib/logstash/outputs/s3.rb (hunk 1 above)
Comment (mashhurs) [reply lib/logstash/outputs/s3.rb:275]: Unfortunately, we cannot use `File.zero` here. `file` is a `TemporaryFile` instance which carries internally file descriptor (FD). When calculating the size, it has an internal calculation strategy that aligns on FD or calls directly `File#size` if FD is closed.

> they might be deleted by the upload_file method before reaching this size

Your concern makes sense for the worst case scenario. Upload is an async operation here and there is no guarantee when it finishes before size check or later. So far, I have been testing multiple times didn't face but for the safety adding `Dir.exist?(file.temp_path)` would be better.

File: spec/outputs/s3_spec.rb (hunk 2 above)
Comment (mashhurs) [reply spec/outputs/s3_spec.rb:276]: Yeah, from the best practise I also usually avoid accessing internal properties but for our case I do believe we save resource cost and we have a single isolated method with periodic checker, will update with required changes.

File: lib/logstash/outputs/s3.rb (hunk 1 above)
Comment (edmocosta) [reply lib/logstash/outputs/s3.rb]: Just to clarify, as I don't know this plugin well, Isn't the `temp_path` the same for all temporary files? If it's, I'd let it as it was, and just `rescue` failures, so other files on the loop would have the chance to get uploaded if one fail. for example:

``` ruby
      ...
      begin
        clean_temporary_file(file) if file.size == 0
      rescue => e
        @logger.debug("Failed removing temporary file", :path => file.path, :exception => e.class, :message => e.message)
      end
      ...
```

Otherwise, LGTM! :)

Comment (mashhurs) [reply lib/logstash/outputs/s3.rb]: > Just to clarify, as I don't know this plugin well, Isn't the temp_path the same for all temporary files?

This plugin has alot magics. No, every temp path is unique.

**Review by edmocosta — APPROVED** [review]
LGTM
